### DISCUSSION TITLE
Wasm build fails

### DISCUSSION THREAD

**Discussion by fatichar** [unoplatform/Uno.Wasm.Bootstrap#657] [Q&A] [answered]
I get this error when I build my WASM project:
```
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.8103018] Unhandled exception. System.Exception: Could not resolve Microsoft.Extensions.Configuration.FileExtensions
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.8103483]    at Driver.Resolve(String asm_name, AssemblyKind& kind) in D:\a\1\s\src\Uno.Wasm.Packager\packager.cs:line 242
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.8103571]    at Driver.Import(String ra, AssemblyKind kind) in D:\a\1\s\src\Uno.Wasm.Packager\packager.cs:line 310
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.8103587]    at Driver.Import(String ra, AssemblyKind kind) in D:\a\1\s\src\Uno.Wasm.Packager\packager.cs:line 310
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.8103594]    at Driver.Run(String[] args) in D:\a\1\s\src\Uno.Wasm.Packager\packager.cs:line 726
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : [00:00:00.8103835]    at Driver.Main(String[] args) in D:\a\1\s\src\Uno.Wasm.Packager\packager.cs:line 370
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error : System.Exception: Failed to generate wasm layout (More details are available in diagnostics mode or using the MSBuild /bl switch)
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_v667976303edda96613ed27b40cdf1ccde7b38a60.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 723
5>C:\Users\pravi\.nuget\packages\uno.wasm.bootstrap\7.0.11\build\Uno.Wasm.Bootstrap.targets(207,3): error :    at Uno.Wasm.Bootstrap.ShellTask_v667976303edda96613ed27b40cdf1ccde7b38a60.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 255
```

**Comment by navio-n** [comment]
Also:

Putting MSBuild project build output verbosity to Normal:

1>  [00:00:00.5320249] Done resolving assemblies
1>  Running [E:\User\repos\Uno.Samples\UI\SkiaSharpTest\SkiaSharpTest\SkiaSharpTest.Wasm\obj\Debug\net5.0\workAot]: cmd /c "C:\Users\User\.uno\emsdk\emsdk-3.1.12\emsdk\emsdk_env.bat 2>&1 && C:\Users\User\AppData\Local\Temp\dotnet-runtime-wasm-windows-633c7b6-0cee4aaff9c-3338312971-Release\tools\ninja.exe "
1>  [00:00:00.0062749] The system cannot find the path specified.

Only 
.uno-install-done file is found in
C:\Users\User\.uno\emsdk\emsdk-3.1.12\

There is no emsdk folder including emsdk_env.bat

**Comment by navio-n** [comment]
Using the latest versions of Uno and Skiasharp

**Comment by jeromelaban** [comment, marked as answer]
@fatichar this generally means that there's a discrepancy in referenced packages. If you can provide a repro project, it'll help determine which one that is.

@navio-n if you're getting this error, it's likely that emscripten did not get installed properly. It's generally caused by `git` or `python` missing on your system. Try deleting `C:\Users\User.uno\emsdk` and build again.

**Reply by navio-n** [reply]
I installed more times Python (both 3.9 and 3.10) from Windows Store.

But uno-check is not seeing it. Only that error occurs when I run it.  I do not know why?

I installed https://git-scm.com/download/win

This only happens when I add Skiasharp dependencies to Wasm

**Reply by jeromelaban** [reply]
The python validation is not working properly in some cases, we're still validating.

Can you [provide a binlog](https://platform.uno/docs/articles/uno-builds-troubleshooting.html#generating-msbuild-binary-log-files) so we can see further into the issue?

**Reply by navio-n** [reply]
[msbuild.zip](https://github.com/unoplatform/Uno.Wasm.Bootstrap/files/10376018/msbuild.zip)

**Reply by jeromelaban** [reply]
Thanks. Can you validate that there's a file named `C:\Users\User.uno\emsdk\emsdk-3.1.12\emsdk\emsdk_env.bat` ?

**Reply by navio-n** [reply]
No. See folder in picture:

![image](https://user-images.githubusercontent.com/25226756/211655102-8f956ec4-6d71-414d-9043-a65aa6fe22e7.png)

**Reply by jeromelaban** [reply]
Thanks. This means that for some reason either git and/or python are not installed or useable. Can you delete the emsdk folder, then run a new build with a binlog? It will log the results of the emsdk installation.

The interesting part is that error messages should have been raising that issue, but it's clear that there's a missing path.

**Reply by navio-n** [reply]
Strange. Now there is the folder and .bat file:
C:\Users\User\.uno\emsdk\emsdk-3.1.12\emsdk.bat
\ between User and .uno. Maybe you forgot that \
[msbuild.zip](https://github.com/unoplatform/Uno.Wasm.Bootstrap/files/10386768/msbuild.zip)


But the building still fails. See binlog

**Reply by jeromelaban** [reply]
Ok, so for some reason, this time, it's more about emscripten that failed to install, because the download failed:
```
[00:00:02.6933982] Error: Downloading URL 'https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/node-v14.18.2-win-x64.zip': <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)>
```

Are you able to download that file manually to see if it works? Otherwise, this step needs to success for eveything to work (you'll still need to remove the emsdk folder for each try). This also means that the bootstrapper is not validating the error code of the emscripten setup properly, let's track in https://github.com/unoplatform/Uno.Wasm.Bootstrap/issues/658.

**Reply by navio-n** [reply]
Where to put/extract that file?

**Reply by jeromelaban** [reply]
You can't it needs to be downloaded by python. The question here is that Python, for some reason, is not able to download this file because of an SSL issue.

It looks like this issue:

https://github.com/emscripten-core/emscripten/issues/9036#issuecomment-733412647

Can you try running what's explained there?

**Reply by navio-n** [reply]
Tried but the same error

**Reply by jeromelaban** [reply]
Then I do not have a workaround at this time. It's a python related error that we cannot troubleshoot on our end. If you find a way to update the python certificate store for your installation, it will likely fix this installation error.

**Reply by navio-n** [reply]
Thank you! Will try tomorrow on another computer.

**Reply by navio-n** [reply]
Hello Jerome,

I tried on a different computer but the same error.  Also, I tried creating a new Uno Platform Wasm Hello World application and it is building properly on both computers but as soon as I put in dependencies build fails:

	  <PackageReference Include="SkiaSharp.NativeAssets.WebAssembly" Version="2.88.3" />
	  <PackageReference Include="SkiaSharp.Views.Uno" Version="2.88.3" />

Do I need to put something else in Linker or Application config to add SkiaSharp?

**Reply by jeromelaban** [reply]
Can you provide the binlog for the new computer as well?

**Reply by navio-n** [reply]
Two logs. The first is OK without Skia and Failed after Skia Dependencies were added on another computer.

[msbuild_Fail_SkiaAdded.zip](https://github.com/unoplatform/Uno.Wasm.Bootstrap/files/10449040/msbuild_Fail_SkiaAdded.zip)
[msbuild_OK_NoSkia.zip](https://github.com/unoplatform/Uno.Wasm.Bootstrap/files/10449041/msbuild_OK_NoSkia.zip)

**Reply by jeromelaban** [reply]
Thanks for the logs. It's the same issue indeed. Can you do the following:
- Delete the `C:\Users\nikol\.uno\emsdk` folder
- Immediately after, without any other wasm build happening, build your project again with skia enabled and binlog enabled

Post the resulting log. Thanks!

**Reply by navio-n** [reply]
On this computer, this solved the problem. Now it is built.

emsdk-2.0.23 is now cca 1.3GB

On old computer where it is not working is only 2.5MB.

Strange

**Reply by jeromelaban** [reply]
In your other computer, you're still getting the python error that is related to SSL certificates and file downloads. Maybe uninstalling Python and reinstalling it could help?

**Reply by navio-n** [reply]
Tried to reinstall but then figured out that the Inkscape SVG program installed in his folder was some lower version of Python. Now it is working on both computers!

**Reply by jeromelaban** [reply]
Very good, thanks for the updates!
